curated_transformers/models/attention.py

```diff
@@ -393,74 +269,77 @@ def forward(
             positions - (batch, seq_len)
         """
 
-        if self.qkv_mode == QkvMode.SEPARATE:
-            q = self.query(x)
-            k = self.key(x)
-            v = self.value(x)
-
-            q = split_heads(q, self.num_heads)
-            k = split_heads(k, self.num_heads)
-            v = split_heads(v, self.num_heads)
-        elif self.qkv_mode == QkvMode.MERGED_SPLIT_BEFORE:
-            proj = self.input(x)
-            proj = split_heads(proj, self.num_heads)
-            q, k, v = proj.chunk(3, dim=-1)
-        else:
-            proj = self.input(x)
-            q, k, v = proj.chunk(3, dim=-1)
-
-            k = split_heads(k, self.num_heads)
-            q = split_heads(q, self.num_heads)
-            v = split_heads(v, self.num_heads)
+        query, key, value = self._query_key_value(x)
 
-        dims_per_head = self.dims_per_head
-        rotary_dims = self.rotary_dims
-
-        if rotary_dims == dims_per_head:
-            # Fast path: we apply rotary embeddings the full key/query vectors.
-            k = self.rotary_embeds(k)
-            q = self.rotary_embeds(q)
-        else:
-            # Otherwise, split up key/query vectors, apply rotary embeddings
-            # and concatenate again.
-            k_rotary, k_rest = k.split([rotary_dims, dims_per_head - rotary_dims], -1)
-            q_rotary, q_rest = q.split([rotary_dims, dims_per_head - rotary_dims], -1)
-
-            # If a cache was provided, but no positions, assume that the
-            # positions of the current batch continue from the cache.
-            if cache is not None and positions is None:
-                cache_len = cache.key.size(-2)
-                seq_len = k.size(-2)
-                positions = torch.arange(
-                    cache_len,
-                    cache_len + seq_len,
-                    dtype=torch.long,  # `torch.int32` isn't supported in indexing operations prior in torch<2.0.0.
-                    device=k_rotary.device,
-                ).repeat(x.size(0), 1)
-
-            # Apply rotary embeddings.
-            k_rotary = self.rotary_embeds(k_rotary, positions=positions)
-            q_rotary = self.rotary_embeds(q_rotary, positions=positions)
-
-            q = torch.cat([q_rotary, q_rest], dim=-1)
-            k = torch.cat([k_rotary, k_rest], dim=-1)
+        if hasattr(self, "rotary_embeds"):
```

Comment:
I think we can set `self.rotary_embeds` to `None` if the config is not provided and use a `None` check here.

Reply:
Yeah, I was doubting what to do, since the constructor is a little longer when making it an `Optional`, but it is probably clearer, so changed.